### PR TITLE
feat(Carta Giovani Nazionale): [#176959318,#177244438,#177244411] Add OTP generation network / actions / saga / reducer

### PR DESCRIPTION
**Pull request by Undermaken**
## Short description
This PR adds all stuff need to handle otp generation

- api networking
- actions
- saga
- reducer

Schermata 2021-03-09 alle 15 04 42

### how to test
- make sure to have the last commit of `io-dev-server `(`master` branch)

### CODE REVIEW

File: ts/features/bonus/cgn/store/actions/otp.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import { ActionType, createAsyncAction } from "typesafe-actions";
+import { NetworkError } from "../../../../../utils/errors";
+import { Otp } from "../../../../../../definitions/cgn/Otp";
+
+/**
+ * handle CGN Otp generation
+ */
+export const cngGenerateOtp = createAsyncAction(
```
Comment (CrisTofani) [inline ts/features/bonus/cgn/store/actions/otp.ts:8]:
```suggestion
export const cgnGenerateOtp = createAsyncAction(
```

File: ts/features/bonus/cgn/store/actions/otp.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  "CGN_GENERATE_OTP_REQUEST",
+  "CGN_GENERATE_OTP_SUCCESS",
+  "CGN_GENERATE_OTP_FAILURE"
+)<void, Otp, NetworkError>();
+
+export type CgnOtpActions = ActionType<typeof cngGenerateOtp>;
```
Comment (CrisTofani) [inline ts/features/bonus/cgn/store/actions/otp.ts:14]:
```suggestion
export type CgnOtpActions = ActionType<typeof cgnGenerateOtp>;
```

File: ts/features/bonus/cgn/store/reducers/otp.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+const INITIAL_STATE: CgnOtpState = {
+  data: remoteUndefined
+};
+
+const reducer = (
+  state: CgnOtpState = INITIAL_STATE,
+  action: Action
+): CgnOtpState => {
+  switch (action.type) {
+    case getType(cngGenerateOtp.request):
```
Comment (CrisTofani) [inline ts/features/bonus/cgn/store/reducers/otp.ts:28]:
```suggestion
    case getType(cgnGenerateOtp.request):
```

File: ts/features/bonus/cgn/store/reducers/otp.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+      return {
+        ...state,
+        data: remoteLoading
+      };
+    case getType(cngGenerateOtp.success):
+      return {
+        ...state,
+        data: remoteReady(action.payload)
+      };
+    case getType(cngGenerateOtp.failure):
```
Comment (CrisTofani) [inline ts/features/bonus/cgn/store/reducers/otp.ts:38]:
```suggestion
    case getType(cgnGenerateOtp.failure):
```

**Review by CrisTofani — CHANGES_REQUESTED** [review]
Just find out some typos, on the rest lgtm

File: ts/features/bonus/cgn/store/reducers/otp.ts (hunk 4 above)
Comment (Undermaken) [reply ts/features/bonus/cgn/store/reducers/otp.ts]: thanks!
changed in https://github.com/pagopa/io-app/pull/2882/commits/0c7f2ce791d125d60b89fc83901c40150c741946

File: ts/features/bonus/cgn/store/reducers/otp.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+  state: CgnOtpState = INITIAL_STATE,
+  action: Action
+): CgnOtpState => {
+  switch (action.type) {
+    case getType(cngGenerateOtp.request):
+      return {
+        ...state,
+        data: remoteLoading
+      };
+    case getType(cngGenerateOtp.success):
```
[the remark this answers is not in the document]
Comment (Undermaken) [reply ts/features/bonus/cgn/store/reducers/otp.ts]: thanks!
changed in https://github.com/pagopa/io-app/pull/2882/commits/0c7f2ce791d125d60b89fc83901c40150c741946

File: ts/features/bonus/cgn/store/reducers/otp.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
+import { getType } from "typesafe-actions";
+import { Action } from "../../../../../store/actions/types";
+import { GlobalState } from "../../../../../store/reducers/types";
+import { NetworkError } from "../../../../../utils/errors";
+import { Otp } from "../../../../../../definitions/cgn/Otp";
+import { cngGenerateOtp } from "../actions/otp";
```
[the remark this answers is not in the document]
Comment (Undermaken) [reply ts/features/bonus/cgn/store/reducers/otp.ts]: thanks!
changed in https://github.com/pagopa/io-app/pull/2882/commits/0c7f2ce791d125d60b89fc83901c40150c741946

File: ts/features/bonus/cgn/store/reducers/otp.ts (hunk 3 above)
Comment (Undermaken) [reply ts/features/bonus/cgn/store/reducers/otp.ts]: thanks!
changed in https://github.com/pagopa/io-app/pull/2882/commits/0c7f2ce791d125d60b89fc83901c40150c741946

File: ts/features/bonus/cgn/store/actions/otp.ts (hunk 2 above)
Comment (Undermaken) [reply ts/features/bonus/cgn/store/actions/otp.ts]: thanks!
changed in https://github.com/pagopa/io-app/pull/2882/commits/0c7f2ce791d125d60b89fc83901c40150c741946

File: ts/features/bonus/cgn/store/actions/otp.ts (hunk 1 above)
Comment (Undermaken) [reply ts/features/bonus/cgn/store/actions/otp.ts]: thanks!
changed in https://github.com/pagopa/io-app/pull/2882/commits/0c7f2ce791d125d60b89fc83901c40150c741946

**Pull request merged**